Temenos BCNMs

cocosolos opened the pull request:
All 8 level 75 Temenos zones.

<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

cocosolos committed:
Temenos East

cocosolos committed:
Temenos North

cocosolos committed:
Temenos West

cocosolos committed:
Temenos Central 1st floor

cocosolos committed:
Temenos Central 2nd floor

cocosolos committed:
Temenos Central 3rd floor

cocosolos committed:
Temenos Central Basement

cocosolos committed:
Temenos Central 4th floor

cocosolos committed:
Update to use noKiller

ibm2431 commented on scripts/zones/Temenos/mobs/Air_Elemental.lua:
```diff
@@ -2,37 +2,71 @@
+                [ID.mob.TEMENOS_C_MOB[2]+5] = function()
+                    GetMobByID(ID.mob.TEMENOS_C_MOB[2]):setMod(tpz.mod.WINDDEF, -128)
+                    if GetMobByID(ID.mob.TEMENOS_C_MOB[2]+6):isAlive() then
+                        DespawnMob(ID.mob.TEMENOS_C_MOB[2]+6)
```
Could shorten this switch here (and other non-dark elementals) a little by:

```lua
if mobID >= ID.mob.TEMENOS_C_MOB[2] then
    -- Central
else
    local crateID = ID.npc.TEMENOS_E_CRATE[3] + (mobID - ID.mob.TEMENOS_E_MOB[3])
    GetNPCByID(crateID):setPos(mobX, mobY, mobZ)
    tpz.limbus.spawnRandomCrate(crateID, player, "crateMaskF3", battlefield:getLocalVar("crateMaskF3"), true)
end
```

ibm2431 commented on scripts/zones/Temenos/mobs/Mystic_Avatar.lua:
```diff
@@ -2,87 +2,106 @@
+            local crateMask = battlefield:getLocalVar("crateMaskF6")
+            if crateMask == 0 then
+                tpz.limbus.handleDoors(battlefield, true, ID.npc.TEMENOS_E_GATE[6])
+            end
```
```lua
if mobID <= ID.mob.TEMENOS_E_MOB[6] + 4 then
    local floor = ((mobID - (ID.mob.TEMENOS_E_MOB[1] + 4)) / 9) + 1
    local crateMask = battlefield:getLocalVar("crateMaskF" .. floor)
    if crateMask == 0 then
        tpz.limbus.handleDoors(battlefield, true, ID.npc.TEMENOS_E_GATE[floor])
    end
else
    -- Central Temenos
end
```

ibm2431 commented on scripts/zones/Temenos/mobs/Mystic_Avatar.lua:
```diff
@@ -2,87 +2,106 @@
+            GetMobByID(ID.mob.TEMENOS_C_MOB[2]):setMod(tpz.mod.WATERDEF, -128)
+            if GetMobByID(ID.mob.TEMENOS_C_MOB[2]+3):isAlive() then
+                DespawnMob(ID.mob.TEMENOS_C_MOB[2]+3)
+                SpawnMob(ID.mob.TEMENOS_C_MOB[2]+9)
```
```lua
elseif mobID >= ID.mob.TEMENOS_C_MOB[2]+9 then
    local element_offset = mobID - ID.mob.TEMENOS_C_MOB[2]+8
    local partner_offset = element_offset % 6 -- Levithan's partner starts at 0
    GetMobByID(ID.mob.TEMENOS_C_MOB[2]):setMod(tpz.mod.FIREDEF - 1 + element_offset)
    if GetMobByID(ID.mob.TEMENOS_C_MOB[2] + 3 + partner_offset):isAlive() then
        DespawnMob(ID.mob.TEMENOS_C_MOB[2] + 3 + partner_offset)
        SpawnMob(ID.mob.TEMENOS_C_MOB[2] + 9 + partner_offset)
    end
end
```

ibm2431 reviewed (commented):
Review on Eastern Tower as of 0d309b0

(And by extension, Central 2F as of eb4b878)

Regarding Carbuncle's decreasing resistances:
> [Note: Carbuncle has exceptionally high damage taken resistance upon first entering the zone; this is progressively weakened by defeating the six elementals/avatars scattered around the zone.](https://www.bg-wiki.com/bg/Central_Temenos_-_2nd_Floor)

Is this elemental resistances as coded, or flat physical/magical resistances?

ibm2431 commented on scripts/zones/Temenos/mobs/Goblin_Slaughterman.lua:
```diff
@@ -2,20 +2,28 @@
-      GetNPCByID(16928770+450):setStatus(tpz.status.NORMAL);
- end
-end;
+        if mobID == ID.mob.TEMENOS_N_MOB[1]+1 and random % 2 == 1 then
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[1]):setStatus(tpz.status.NORMAL)
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[1]+1):setStatus(tpz.status.NORMAL)
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[1]+2):setStatus(tpz.status.NORMAL)
+        elseif mobID == ID.mob.TEMENOS_N_MOB[1] and random % 2 == 0 then
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[1]):setStatus(tpz.status.NORMAL)
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[1]+1):setStatus(tpz.status.NORMAL)
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[1]+2):setStatus(tpz.status.NORMAL)
```
Can combine these with the random modulo into one:
```lua
if mobID == ID.mob.TEMENOS_N_MOB[1] + (random % 2) then
    -- Chests
end
```

Although, there is a ffxiclopedia and bgwiki conflict:
> [Chests appear after defeating both Goblin Slaughterman.](https://www.bg-wiki.com/bg/Northern_Temenos)

> [The chest dropper is one of the Goblin Slaughterman in the NW corner OR SE corner. ](https://ffxiclopedia.fandom.com/wiki/Temenos_-_Northern_Tower)

cocosolos replied:
Yeah I'll change this, everything I've seen points to either both of them needing to die or just the top one. I can't find any video of someone skipping the first Slaughterman and killing the upstairs one first.

cocosolos commented:
> Regarding Carbuncle's decreasing resistances:
> 
> > [Note: Carbuncle has exceptionally high damage taken resistance upon first entering the zone; this is progressively weakened by defeating the six elementals/avatars scattered around the zone.](https://www.bg-wiki.com/bg/Central_Temenos_-_2nd_Floor)
> 
> Is this elemental resistances as coded, or flat physical/magical resistances?

This was already being done with elemental resistance in the Mystic Avatar script, so I just moved it to the elementals. It sounds like it should be damage taken though not elemental resist.

ibm2431 commented on scripts/zones/Temenos/mobs/Telchines_White_Mage.lua:
```diff
@@ -2,20 +2,25 @@
-    GetNPCByID(16928768+211):setStatus(tpz.status.NORMAL);
- end
-end;
+        if random % 2 == 1 then
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[3]):setStatus(tpz.status.NORMAL)
```
bgwiki says "and" the BRD, ffxiclopedia says "and/or". We'll have to look into this for later.

cocosolos replied:
This looks like the first floor to me, either both of them or just the far one. Same situation, all videos have chests spawning on the BRD after already killing WHM.

ibm2431 commented on scripts/zones/Temenos/mobs/Telchines_Monk.lua:
```diff
@@ -2,20 +2,45 @@
+        local battlefield = player:getBattlefield()
+        local random = battlefield:getLocalVar("randomF3")
+
+        if mobID - ID.mob.TEMENOS_N_MOB[3] == random - 1 then
```
Is this a `random == 5` case?

cocosolos replied:
I'm not sure what I did here, but it looks wrong. `random` can be 1-6, but doors won't open if it's 1 or 2. Will fix.

Correction: 1 and 2 is handled by the BRD or WHM that doesn't drop the chests.

ibm2431 commented on scripts/zones/Temenos/mobs/Praetorian_Guard_CCCXI.lua:
```diff
@@ -2,21 +2,22 @@
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[5]+1):setStatus(tpz.status.NORMAL)
+            GetNPCByID(ID.npc.TEMENOS_N_CRATE[5]+2):setStatus(tpz.status.NORMAL)
+        end
+        if GetNPCByID(ID.npc.TEMENOS_N_GATE[5]):getAnimation() == tpz.animation.CLOSE_DOOR then
```
Do the doors on the 5th floor open after _any_ kill, or a random one (like previous floors)?

cocosolos replied:
Every video I've seen door opens after first kill.

ibm2431 commented on scripts/zones/Temenos/mobs/Cryptonberry_Abductor.lua:
```diff
@@ -0,0 +1,51 @@
+
+function onMobDeath(mob, player, isKiller)
+    if isKiller then
+        if GetNPCByID(ID.npc.TEMENOS_N_GATE[6]):getAnimation() == tpz.animation.CLOSE_DOOR then
```
Same question on the _any_ Cryptonberry

cocosolos replied:
Think this is same as the Antican.
From wiki:
`Either the Cryptonberry Designator or Cryptonberry Abductor will open the portal. The lone THF is recommended.`
Recommending the THF implies a choice from any of them.

ibm2431 commented on scripts/zones/Temenos/bcnms/temenos_northern_tower.lua:
```diff
@@ -2,38 +2,47 @@
+    SetServerVariable("[Temenos_Northern_Tower]Time", battlefield:getTimeLimit()/60)
+    tpz.limbus.handleDoors(battlefield)
+    tpz.limbus.setupArmouryCrates(battlefield:getID())
+    DespawnMob(ID.mob.TEMENOS_N_MOB[3]+3)
+    DespawnMob(ID.mob.TEMENOS_N_MOB[4]+3)
+    DespawnMob(ID.mob.TEMENOS_N_MOB[6]+4)
+    DespawnMob(ID.mob.TEMENOS_N_MOB[6]+8)
+    DespawnMob(ID.mob.TEMENOS_N_MOB[6]+12)
```
What are these despawns for?

cocosolos replied:
I think the issue was I was getting navmesh errors if I don't spawn the pets at the start, and they'll be spawned before they should be (DRG pets) if I don't despawn them.

ibm2431 reviewed (commented):
Review on North Tower as of 88a843e

TeoTwawki commented:
> > Regarding Carbuncle's decreasing resistances:
> > > [Note: Carbuncle has exceptionally high damage taken resistance upon first entering the zone; this is progressively weakened by defeating the six elementals/avatars scattered around the zone.](https://www.bg-wiki.com/bg/Central_Temenos_-_2nd_Floor)
> > 
> > 
> > Is this elemental resistances as coded, or flat physical/magical resistances?
> 
> This was already being done with elemental resistance in the Mystic Avatar script, so I just moved it to the elementals. It sounds like it should be damage taken though not elemental resist.

Probably the element defense mods, which act like damage taken mods of 1 specified element each - the early revisions often mixed these up with resistance rate mods

cocosolos committed:
Clean up Temenos East mobs

ibm2431 commented on scripts/zones/Temenos/mobs/Enhanced_Vulture.lua:
```diff
@@ -2,27 +2,24 @@
-        GetMobByID(16928964):updateEnmity(target);
-end;
+function onMobEngaged(mob, target)
+    GetMobByID(ID.mob.TEMENOS_W_MOB[7]):updateEnmity(target)
+    GetMobByID(ID.mob.TEMENOS_W_MOB[7]+1):updateEnmity(target)
+    GetMobByID(ID.mob.TEMENOS_W_MOB[7]+2):updateEnmity(target)
+    GetMobByID(ID.mob.TEMENOS_W_MOB[7]+3):updateEnmity(target)
+    GetMobByID(ID.mob.TEMENOS_W_MOB[7]+4):updateEnmity(target)
+    GetMobByID(ID.mob.TEMENOS_W_MOB[7]+5):updateEnmity(target)
```
I probably should have asked earlier- and I know this was already here - but can we use standard linking mechanics for these situations?

cocosolos replied:
Since they link by sight you'd be able to pull from behind 1 at a time. I think they all attack at once, but may be wrong.

cocosolos committed:
Change some mechanics in Temenos North

Have chests spawn after both Slaughterman are killed on floor 1.
Have chests spawn after both WHM and BRD are killed on floor 3.
Portal opens from random Dustman, MNK, or DRG.

ibm2431 reviewed (commented):
Review on Temenos West as of ab2eef9

ibm2431 commented on scripts/zones/Temenos/mobs/Airi.lua:
```diff
@@ -2,24 +2,26 @@
-  end
-end;
+function onMobEngaged(mob, target)
+    if GetMobByID(ID.mob.TEMENOS_C_MOB[1]+1):isDead() then
+        mob:addStatusEffect(tpz.effect.REGAIN, 7, 3, 0)
+        mob:addStatusEffect(tpz.effect.REGEN, 50, 3, 0)
```
Usual "this was already here, but...." disclaimer: What happens if a party decides to engage one, hold it, and then kill the linked mob to try to get around the "strength increasing" effect?

(Or a wipe after killing the linked mob.)

We may want to move the "gets stronger" logic to the onDeath of the linked mob.

cocosolos replied:
Yeah that sounds better. I think it's supposed to force a textless 2hr animation too, not sure how to do that.

ibm2431 replied:
> supposed to force a textless 2hr animation too, not sure how to do that.

@59blargedy Time for (Vulture 2)? 😉

59blargedy replied:
i think substitute (id 307) is more commonly used for messageless 2 hour animtion
cat 11 id 307 animation 439 message 0
(the vulture 2 and 3 thing from sea are weird?!?!)

ibm2431 reviewed (commented):
Review on Central 1F as of 03a8f44

ibm2431 commented on scripts/zones/Temenos/bcnms/central_temenos_3rd_floor.lua:
```diff
@@ -2,37 +2,43 @@
+    SetServerVariable("[Central_Temenos_3rd_Floor]Time", battlefield:getTimeLimit()/60)
+    tpz.limbus.handleDoors(battlefield)
+    tpz.limbus.setupArmouryCrates(battlefield:getID())
+    DespawnMob(ID.mob.TEMENOS_C_MOB[3]+3)
+    DespawnMob(ID.mob.TEMENOS_C_MOB[3]+11)
```
To show I _have_ been paying attention: These are for the Mook Orc DRG and the Boss Yagudo SMN, right?

cocosolos replied:
Yes

ibm2431 commented on scripts/zones/Temenos/mobs/Abyssdweller_Jhabdebb.lua:
```diff
@@ -2,31 +2,34 @@
-  GetMobByID(16929006):updateEnmity(target);
-  GetMobByID(16929007):updateEnmity(target);
-end;
+function onMobEngaged(mob, target)
+    if GetMobByID(ID.mob.TEMENOS_C_MOB[3]+5):isDead() and GetMobByID(ID.mob.TEMENOS_C_MOB[3]+6):isDead() and
+        GetMobByID(ID.mob.TEMENOS_C_MOB[3]+7):isDead() and GetMobByID(ID.mob.TEMENOS_C_MOB[3]+8):isDead() and
+        GetMobByID(ID.mob.TEMENOS_C_MOB[3]+9):isDead() and GetMobByID(ID.mob.TEMENOS_C_MOB[3]+10):isDead()
+    then
+        mob:setMod(tpz.mod.SLASHRES, 1400)
+        mob:setMod(tpz.mod.PIERCERES, 1400)
```
Same as before regarding strength/weaken and engaged v death, but this time to the players' detriment. 😄 

(In this case, being aggroed by a boss before the last mook is dead, and then the boss not getting weaker until it re-engages.)

ibm2431 reviewed (commented):
Review on Central 3F as of 11c55fe

(Central Tower is certainly quicker to review~)

ibm2431 commented on scripts/zones/Temenos/mobs/Temenos_Aern.lua:
```diff
@@ -2,52 +2,69 @@
+                tpz.limbus.extendTimeLimit(battlefield, 5, tpz.zone.TEMENOS)
+            end,
+            [ID.mob.TEMENOS_C_MOB[5]+30] = function()
+                tpz.limbus.extendTimeLimit(battlefield, 5, tpz.zone.TEMENOS)
```
bgwiki is claiming "random Aern"

> [Time limit is 15 minutes, but killing random Temenos Aerns will give time extensions.](https://www.bg-wiki.com/bg/Central_Temenos_-_Basement)

cocosolos replied:
I believe it's 1 random Aern in 4 set groups of Aerns.  I currently have it set as the center mob in each of these groups, so does need to be randomized a bit. They also need to have their loot removed from db and manually drop coins based on how many times they RR.

ibm2431 reviewed (commented):
Review on Temenos Basement as of f83fc92

(I love reviewing Temenos Central Tower!)

ibm2431 commented on scripts/globals/mobskills/citadel_buster.lua:
```diff
@@ -0,0 +1,89 @@
+        for _, member in pairs(players) do
+            member:messageSpecial(ID.text.CITADEL_BASE)
+        end
+        mob:timer(10000, function(mob)
+            for _, member in pairs(players) do
+                member:messageSpecial(ID.text.CITADEL_BASE+1)
+            end
+            mob:timer(10000, function(mob)
+                for _, member in pairs(players) do
+                    member:messageSpecial(ID.text.CITADEL_BASE+2)
+                end
+                mob:timer(5000, function(mob)
+                    for _, member in pairs(players) do
+                        member:messageSpecial(ID.text.CITADEL_BASE+3)
```
Others: I started typing an idea, then spoke with cocosolos on Discord. We might make use of onBattlefieldTick.

ibm2431 commented on scripts/globals/mobskills/cryo_jet.lua:
```diff
@@ -35,6 +35,9 @@ function onMobWeaponSkill(target, mob, skill)
 
     target:takeDamage(dmg, mob, tpz.attackType.MAGICAL, tpz.damageType.ICE)
-
+    if target:hasStatusEffect(tpz.effect.ELEMENTALRES_DOWN) then
+        target:delStatusEffectSilent(tpz.effect.ELEMENTALRES_DOWN)
```
> The follow-up moves after nuclear waste remove elemental resist down?

> im actually not 100% on that

> Then I shall leave a comment for later research

ibm2431 commented on scripts/globals/mobskills/chemical_bomb.lua:
```diff
@@ -33,4 +35,4 @@ function onMobWeaponSkill(target, mob, skill)
 
     -- This likely doesn't behave like retail.
     return typeEffectTwo
```
I know you haven't touched this, but this comment might be addressed as we're about to scrutinize Proto-Ultima.

ibm2431 reviewed (commented):
Review on Central 4th as of f7dc6e4
